File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.java
package iamutkarshtiwari.github.io.ananas.editimage.fragment.paint;

import android.app.Dialog;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import org.jetbrains.annotations.NotNull;

import iamutkarshtiwari.github.io.ananas.BaseActivity;
import iamutkarshtiwari.github.io.ananas.R;
import iamutkarshtiwari.github.io.ananas.editimage.EditImageActivity;
import iamutkarshtiwari.github.io.ananas.editimage.ModuleConfig;
import iamutkarshtiwari.github.io.ananas.editimage.fragment.BaseEditFragment;
import iamutkarshtiwari.github.io.ananas.editimage.fragment.MainMenuFragment;
import iamutkarshtiwari.github.io.ananas.editimage.utils.Matrix3;
import iamutkarshtiwari.github.io.ananas.editimage.view.CustomPaintView;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class PaintFragment extends BaseEditFragment implements View.OnClickListener, BrushConfigDialog.Properties, EraserConfigDialog.Properties {

    public static final int INDEX = ModuleConfig.INDEX_PAINT;
    public static final String TAG = PaintFragment.class.getName();

    private static final float MAX_PERCENT = 100;
    private static final float MAX_ALPHA = 255;
    private static final float INITIAL_WIDTH = 50;

    private boolean isEraser = false;

    //    private CustomPaintView customPaintView;
    private ImageView imageEraser;
    private ImageView imageBrush;

    private BrushConfigDialog brushConfigDialog;
    private EraserConfigDialog eraserConfigDialog;
//    private Dialog loadingDialog;

    private float brushSize = INITIAL_WIDTH;
    private float eraserSize = INITIAL_WIDTH;
    private float brushAlpha = MAX_ALPHA;
    private int brushColor = Color.WHITE;

//    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public static PaintFragment newInstance() {
        return new PaintFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_edit_paint, null);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        view.findViewById(R.id.back_to_main).setOnClickListener(this);

        LinearLayout eraserView = view.findViewById(R.id.eraser_btn);
        eraserView.setOnClickListener(this);
        imageEraser = eraserView.findViewById(R.id.eraser_icon);

        LinearLayout brushView = view.findViewById(R.id.brush_btn);
        brushView.setOnClickListener(this);
        imageBrush = brushView.findViewById(R.id.brush_icon);

        view.findViewById(R.id.settings).setOnClickListener(this);

       /* loadingDialog = BaseActivity.getLoadingDialog(activity, R.string.iamutkarshtiwari_github_io_ananas_loading, false);
        customPaintView = activity.findViewById(R.id.custom_paint_view);*/

        setupOptionsConfig();

        ((PaintToolActionListeners) activity).initPaintView(INITIAL_WIDTH, Color.WHITE, MAX_ALPHA, INITIAL_WIDTH);
//        initStroke();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

    }

    private void setupOptionsConfig() {
        brushConfigDialog = new BrushConfigDialog();
        brushConfigDialog.setPropertiesChangeListener(this);

        eraserConfigDialog = new EraserConfigDialog();
        eraserConfigDialog.setPropertiesChangeListener(this);
    }

    /*private void initStroke() {
        customPaintView.setWidth(INITIAL_WIDTH);
        customPaintView.setColor(Color.WHITE);
        customPaintView.setStrokeAlpha(MAX_ALPHA);
        customPaintView.setEraserStrokeWidth(INITIAL_WIDTH);
    }*/

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.back_to_main) {
            backToMain();
        } else if (view.getId() == R.id.eraser_btn) {
            if (!isEraser) {
                toggleButtons();
            }
        } else if (view.getId() == R.id.brush_btn) {
            if (isEraser) {
                toggleButtons();
            }
        } else if (view.getId() == R.id.settings) {
            showDialog(isEraser ? eraserConfigDialog : brushConfigDialog);
        }
    }

    private void showDialog(BottomSheetDialogFragment dialogFragment) {
        String tag = dialogFragment.getTag();

        // Avoid IllegalStateException "Fragment already added"
        if (dialogFragment.isAdded()) return;

        dialogFragment.show(requireFragmentManager(), tag);

        if (isEraser) {
            ((PaintToolActionListeners) activity).updateEraserData(eraserSize);
//            updateEraserSize();
        } else {
            ((PaintToolActionListeners) activity).updateBrushData(brushSize, brushAlpha, brushColor);
//            updateBrushParams();
        }
    }

    /*@Override
    public void onPause() {
        compositeDisposable.clear();
        super.onPause();
    }*/

    public void backToMain() {
        ((PaintToolActionListeners) activity).handleBackToMain();
        /*customPaintView.reset();
        customPaintView.setVisibility(View.GONE);*/
    }

    public void onShow() {
        activity.mode = EditImageActivity.MODE_PAINT;
        activity.mainImage.setImageBitmap(activity.getMainBit());
        activity.bannerFlipper.showNext();

        ((PaintToolActionListeners) activity).setPaintViewVisibility(VISIBLE);
//        customPaintView.setVisibility(View.VISIBLE);
    }

    private void toggleButtons() {
        isEraser = !isEraser;
        ((PaintToolActionListeners) activity).setIsEraser(isEraser);
//        customPaintView.setEraser(isEraser);
        imageEraser.setImageResource(isEraser ? R.drawable.ic_eraser_enabled : R.drawable.ic_eraser_disabled);
        imageBrush.setImageResource(isEraser ? R.drawable.ic_brush_grey_24dp : R.drawable.ic_brush_white_24dp);
    }

    /*@Override
    public void onDestroy() {
        compositeDisposable.dispose();
        super.onDestroy();
    }*/

    @Override
    public void onColorChanged(int colorCode) {
        brushColor = colorCode;
        ((PaintToolActionListeners) activity).updateBrushData(brushSize, brushAlpha, brushColor);
//        updateBrushParams();
    }

    @Override
    public void onOpacityChanged(int opacity) {
        brushAlpha = (opacity / MAX_PERCENT) * MAX_ALPHA;
        ((PaintToolActionListeners) activity).updateBrushData(brushSize, brushAlpha, brushColor);
//        updateBrushParams();
    }

    @Override
    public void onBrushSizeChanged(int brushSize) {
        if (isEraser) {
            this.eraserSize = brushSize;
            ((PaintToolActionListeners) activity).updateEraserData(eraserSize);
//            updateEraserSize();
        } else {
            this.brushSize = brushSize;
            ((PaintToolActionListeners) activity).updateBrushData(brushSize, brushAlpha, brushColor);
//            updateBrushParams();
        }
    }

    /*private void updateBrushParams() {
        customPaintView.setColor(brushColor);
        customPaintView.setWidth(brushSize);
        customPaintView.setStrokeAlpha(brushAlpha);
    }

    private void updateEraserSize() {
        customPaintView.setEraserStrokeWidth(eraserSize);
    }*/

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        brushConfigDialog = null;
        eraserConfigDialog = null;
    }

    public interface PaintToolActionListeners{
        void initPaintView(float brushSize, int brushColor, float eraserSize, float brushAlpha);
        void updateEraserData(float eraserSize);
        void updateBrushData(float brushSize, float brushAlpha, int brushColor);
        void setIsEraser(boolean isEraser);
        void setPaintViewVisibility(int visibility);
        void resetPaintView();
        void handleBackToMain();
    }
}

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt
package iamutkarshtiwari.github.io.ananas.editimage.fragment.paint

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.fragment.app.activityViewModels
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import iamutkarshtiwari.github.io.ananas.R
import iamutkarshtiwari.github.io.ananas.editimage.EditImageActivity
import iamutkarshtiwari.github.io.ananas.editimage.EditImageActivity.MODE_PAINT
import iamutkarshtiwari.github.io.ananas.editimage.ModuleConfig
import iamutkarshtiwari.github.io.ananas.editimage.fragment.BaseEditFragment
import iamutkarshtiwari.github.io.ananas.editimage.fragment.MainMenuFragment
import iamutkarshtiwari.github.io.ananas.editimage.utils.Matrix3
import iamutkarshtiwari.github.io.ananas.editimage.viewmodel.PaintViewModel
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class PaintFragment : BaseEditFragment(), View.OnClickListener {

    private var imageEraser: ImageView? = null
    private var imageBrush: ImageView? = null

    private val compositeDisposable = CompositeDisposable()

    private val viewModel: PaintViewModel by activityViewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_edit_paint, null)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view.findViewById<View>(R.id.back_to_main).setOnClickListener(this)

        val eraserView = view.findViewById<LinearLayout>(R.id.eraser_btn)
        eraserView.setOnClickListener(this)
        imageEraser = eraserView.findViewById(R.id.eraser_icon)

        val brushView = view.findViewById<LinearLayout>(R.id.brush_btn)
        brushView.setOnClickListener(this)
        imageBrush = brushView.findViewById(R.id.brush_icon)

        view.findViewById<View>(R.id.settings).setOnClickListener(this)
    }

    override fun onClick(view: View) {
        if (view.id == R.id.back_to_main) {
            backToMain()
        } else if (view.id == R.id.eraser_btn) {
            if (viewModel.isEraser.value == false) {
                viewModel.setIsEraser(true)
                toggleButtons()
            }
        } else if (view.id == R.id.brush_btn) {
            if (viewModel.isEraser.value == true) {
                viewModel.setIsEraser(false)
                toggleButtons()
            }
        } else if (view.id == R.id.settings) {

            if (viewModel.isEraser.value == true) {
                showBottomSheet(EraserConfigDialog())
            } else {
                showBottomSheet(BrushConfigDialog())
            }
        }
    }

    private fun showBottomSheet(dialogFragment: BottomSheetDialogFragment) {
        if (dialogFragment.isAdded) return
        dialogFragment.show(parentFragmentManager, dialogFragment.tag)
    }

    override fun backToMain() {
        activityInstance?.apply {
            bottomGallery.currentItem = MainMenuFragment.INDEX
            mainImage.visibility = VISIBLE
            paintView.reset()
            paintView.visibility = View.GONE
            mode = EditImageActivity.MODE_NONE
            bannerFlipper.showPrevious()
        }


        viewModel.resetToDefault()
    }

    override fun onShow() {
        activityInstance?.apply {
            mode = MODE_PAINT
            mainImage.setImageBitmap(mainBit)
            paintView.visibility = VISIBLE
            bannerFlipper.showNext()
        }

    }

    private fun toggleButtons() {
        imageEraser?.setImageResource(
                if (viewModel.isEraser.value == true) R.drawable.ic_eraser_enabled
                else R.drawable.ic_eraser_disabled)
        imageBrush?.setImageResource(
                if (viewModel.isEraser.value == true) R.drawable.ic_brush_grey_24dp
                else R.drawable.ic_brush_white_24dp)
    }

    fun savePaintImage() {
        compositeDisposable.clear()
        activityInstance?.apply {
            val applyPaintDisposable = applyPaint(mainBit)
                    .flatMap { bitmap: Bitmap? ->
                        if (bitmap == null) {
                            return@flatMap Single.error<Bitmap>(Throwable("Error occurred while applying paint"))
                        } else {
                            return@flatMap Single.just(bitmap)
                        }
                    }
                    .subscribeOn(Schedulers.computation())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe { activityInstance?.showLoadingDialog() }
                    .doFinally { activityInstance?.dismissLoadingDialog() }
                    .subscribe { bitmap: Bitmap? ->
                        activityInstance?.paintView?.reset()
                        activityInstance?.changeMainBitmap(bitmap, true)
                        backToMain()
                    }
            compositeDisposable.add(applyPaintDisposable)
        }
    }

    private fun applyPaint(mainBitmap: Bitmap): Single<Bitmap?> {
        return Single.fromCallable {
            var resultBit: Bitmap? = null
            activityInstance?.apply {
                val touchMatrix: Matrix? = this.mainImage?.imageViewMatrix
                resultBit = Bitmap.createBitmap(mainBitmap).copy(Bitmap.Config.ARGB_8888, true)
                resultBit?.apply {
                    val canvas = Canvas(this)
                    val data = FloatArray(9)
                    touchMatrix?.getValues(data)
                    val cal = Matrix3(data)
                    val inverseMatrix = cal.inverseMatrix()
                    val matrix = Matrix()
                    matrix.setValues(inverseMatrix.values)
                    handleImage(canvas, matrix)
                }
            }
            resultBit
        }
    }

    private fun handleImage(canvas: Canvas, matrix: Matrix) {
        val f = FloatArray(9)
        matrix.getValues(f)
        val dx = f[Matrix.MTRANS_X].toInt()
        val dy = f[Matrix.MTRANS_Y].toInt()
        val scale_x = f[Matrix.MSCALE_X]
        val scale_y = f[Matrix.MSCALE_Y]
        canvas.save()
        canvas.translate(dx.toFloat(), dy.toFloat())
        canvas.scale(scale_x, scale_y)
        activityInstance?.paintView?.paintBit?.apply {
            canvas.drawBitmap(this, 0f, 0f, null)
        }
        canvas.restore()
    }

    override fun onPause() {
        compositeDisposable.clear()
        super.onPause()
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
        super.onDestroy()
    }

    companion object {
        const val INDEX = ModuleConfig.INDEX_PAINT
        val TAG = PaintFragment::class.java.name

        @JvmStatic
        fun newInstance(): PaintFragment {
            return PaintFragment()
        }
    }
}